numba/_devicearray.cpp

```diff
@@ -0,0 +1,117 @@
+                                                 /* tp_flags*/
+    "DeviceArray object",                        /* tp_doc */
+    (traverseproc) DeviceArray_traverse,         /* tp_traverse */
+    (inquiry) DeviceArray_clear,                 /* tp_clear */
+    0,                                           /* tp_richcompare */
+    0,                                           /* tp_weaklistoffset */
+    0,                                           /* tp_iter */
+    0,                                           /* tp_iternext */
+    0,                                           /* tp_methods */
+    0,                                           /* tp_members */
+    0,                                           /* tp_getset */
+    0,                                           /* tp_base */
+    0,                                           /* tp_dict */
+    0,                                           /* tp_descr_get */
+    0,                                           /* tp_descr_set */
+    0,                                           /* tp_dictoffset */
+    0,                                           /* tp_init */
+    0,                                           /* tp_alloc */
+    0,                                           /* tp_new */
+    0,                                           /* tp_free */
+    0,                                           /* tp_is_gc */
+    0,                                           /* tp_bases */
+    0,                                           /* tp_mro */
+    0,                                           /* tp_cache */
+    0,                                           /* tp_subclasses */
+    0,                                           /* tp_weaklist */
+    0,                                           /* tp_del */
+    0,                                           /* tp_version_tag */
+    0,                                           /* tp_finalize */
+#if PY_MAJOR_VERSION == 3 && PY_MINOR_VERSION > 7
+    0,                                           /* tp_vectorcall */
+    0,                                           /* tp_print */
+#endif
+};
+
+/* CUDA device array API */
+static void *_DeviceArray_API[1] = {
+    (void*)&DeviceArrayType
+};
+
+MOD_INIT(_devicearray) {
+    PyObject *m, *d;
+    MOD_DEF(m, "_devicearray", "No docs", NULL)
+    if (m == NULL)
+        return MOD_ERROR_VAL;
+
+    PyObject *c_api;
+    c_api = PyCapsule_New((void *)_DeviceArray_API,"_devicearray.c_api", NULL);
+    if (c_api == NULL) {
+        return MOD_ERROR_VAL;
+    }
+
+    DeviceArrayType.tp_new = PyType_GenericNew;
+    if (PyType_Ready(&DeviceArrayType) < 0) {
+        return MOD_ERROR_VAL;
+    }
+    Py_INCREF(&DeviceArrayType);
+    PyModule_AddObject(m, "DeviceArray", (PyObject*)(&DeviceArrayType));
+
+    d = PyModule_GetDict(m);
+    if (d == NULL) {
+        return MOD_ERROR_VAL;
+    }
+
+    PyDict_SetItemString(d, "_DEVICEARRAY_API", c_api);
```

Comment:
The checking is implicit because of the following `if (pyErr_Occurred)` - however, it would probably be better to make the link between what happens in `PyDict_SetItemString` and the return of the error value, so this block would probably be better as:

```C
    int error = PyDict_SetItemString(d, "_DEVICEARRAY_API", c_api);
    Py_DECREF(c_api);
    if (error) {
        return MOD_ERROR_VAL;
```

Reply:
Something similar to the above (but with a goto to the cleanup section) now added.